### PR TITLE
[Excel](Custom Functions) Batching article

### PR DESCRIPTION
**Pull request by mscharlock**
There is a link problem in here that I will fix, I just want to get it out so Linda can look at it as it's new content. :) @lindalu-MSFT would you also be willing to review this? This article is also dependent on code in a separate PR here: https://github.com/OfficeDev/PnP-OfficeAddins/pull/32

### CODE REVIEW

**Comment by o365devx** [comment]
OPS Build status updates of commit _[bf9e7a3](https://github.com/OfficeDev/office-js-docs-pr/pull/712/commits/bf9e7a31dfbaa84272bfc1cc930bbbdd53dff3f4)_: 

### :warning: Validation status: warnings


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/excel/custom-functions-batching.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/excel/custom-functions-batching.md) | :warning:Warning | | [Details](#user-content-docs%2Fexcel%2Fcustom-functions-batching.md)
[docs/excel/custom-functions-changelog.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/excel/custom-functions-changelog.md) | :white_check_mark:Succeeded | |
[docs/images/Batch1.svg](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/images/Batch1.svg) | :white_check_mark:Succeeded | |
[docs/images/Batch2.svg](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/images/Batch2.svg) | :white_check_mark:Succeeded | |
[docs/images/Batch3.svg](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/images/Batch3.svg) | :white_check_mark:Succeeded | |
[docs/images/Batch4.svg](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/images/Batch4.svg) | :white_check_mark:Succeeded | |
[docs/toc.yml](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/toc.yml) | :white_check_mark:Succeeded | |

<a name="docs%2Fexcel%2Fcustom-functions-batching.md"></a>
### [docs/excel/custom-functions-batching.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/excel/custom-functions-batching.md)
 - **Line 12**: **[Warning]** Invalid file link:(~/tutorials/).

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C1%5C28%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201901282129399353-712%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=hkMS0ybyHsGtXrNB9biBT%2BFXISfBMAP4LSDbr1iTNQ0%3D&st=2019-01-28T21%3A25%3A54Z&se=2019-02-28T21%3A30%3A54Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: docs/excel/custom-functions-batching.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+    To give a visual example, the following diagram shows how the information from the `batch` array has been copied into the `batchCopy` array and the stock ticker string information separated out to the `requestBatch` array. The `batch` array has been emptied and could continue to fill with requests in the meantime while the `requestBatch` is processing.
+
+    ![The batch array is copied and emptied.](../images/Batch2.svg "Batching diagram")
+
+    To fill in the third `TODO` item, add the following code.
+
+    ```js
+    function _makeRemoteRequest() {
+        // Copy the shared batch and allow the building of a new batch while you are waiting for a response.
```
Comment (lindalu-MSFT) [inline docs/excel/custom-functions-batching.md]: Copy the shared batch and build a new batch while you await a response.

File: docs/excel/custom-functions-batching.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+            requestBatch[i] = {
+                "operation": batchCopy[i].operation,
+                "arguments": batchCopy[i].arguments
+            };
+    }
+    ```
+
+* **TODO #4**
+
+    In the fourth `TODO`, you'll continue filling in the `_makeRemoteRequest` function.
```
Comment (lindalu-MSFT) [inline docs/excel/custom-functions-batching.md]: To complete the fourth 'TODO', continue to fill in the '_makeRemoteRequest' function.

File: docs/excel/custom-functions-batching.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+    The invocation entries from your `batchCopy` array still have unresolved Promises that need to be fulfilled. When the API returns the information that you can use to fulfill these Promises, you'll match up each response from the API to the Promises in the `batchCopy` array.
```
Comment (lindalu-MSFT) [inline docs/excel/custom-functions-batching.md]: "When the API returns the information you need to fulfill..."
"match up" should be "match"

File: docs/excel/custom-functions-batching.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        // and resolve the invocation Promise with its corresponding response value.
+
+            for (var i = 0; i < responseBatch.length; i++) {
+            batchCopy[i].resolve(responseBatch[i]);
+            }
+        });
+    }
+    ```
+
+    Through this process, the Promise for each invocation is fulfilled and each _pushOperation can be returned in the cell where the function was originally called. You can see in the diagram that follows that the API returned information to the `requestBatch` array, which resolved the Promises in the `batchCopy` array, which then provides the information for the return values of the original custom function invocations. The `batch` array is still empty, but could also begin filling with additional requests for the next batch to be processed.
```
Comment (lindalu-MSFT) [inline docs/excel/custom-functions-batching.md]: "in the following diagram"
"provides" should be "provided"

File: docs/excel/custom-functions-batching.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+    ![The API returns information to complete the unfulfilled Promises.](../images/Batch4.svg "Batching diagram")
+
+    With completion of this step, you're done filling in `TODO` items in this sample.
+
+    The rest of this code sample simulates what a public API would be doing to resolve your API request, for your reference.
+
+## Applying this pattern to your own project
```
Comment (lindalu-MSFT) [inline docs/excel/custom-functions-batching.md]: Apply this pattern to your own project

File: docs/excel/custom-functions-batching.md (hunk 4 above)
Comment (mscharlock) [reply docs/excel/custom-functions-batching.md]: I went with "Through this process, ...each is fulfilled. Each _pushOp returns" and then changed all the other tenses to present.

**Comment by o365devx** [comment]
OPS Build status updates of commit _[69a91b9](https://github.com/OfficeDev/office-js-docs-pr/pull/712/commits/69a91b90c5a9e32f9785ba0913d6d26c4cdde206)_: 

:clock10: **Preparing**: average preparing time is _44 sec(s)_

**Comment by o365devx** [comment]
OPS Build status updates of commit _[69a91b9](https://github.com/OfficeDev/office-js-docs-pr/pull/712/commits/69a91b90c5a9e32f9785ba0913d6d26c4cdde206)_: 

:clock10: **Preparing**: average preparing time is _44 sec(s)_

**Comment by o365devx** [comment]
OPS Build status updates of commit _[69a91b9](https://github.com/OfficeDev/office-js-docs-pr/pull/712/commits/69a91b90c5a9e32f9785ba0913d6d26c4cdde206)_: 

:clock10: **Incremental building**: average incremental building time is _29 sec(s)_

**Comment by o365devx** [comment]
OPS Build status updates of commit _[69a91b9](https://github.com/OfficeDev/office-js-docs-pr/pull/712/commits/69a91b90c5a9e32f9785ba0913d6d26c4cdde206)_: 

### :warning: Validation status: warnings


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/excel/custom-functions-batching.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/excel/custom-functions-batching.md) | :warning:Warning | | [Details](#user-content-docs%2Fexcel%2Fcustom-functions-batching.md)
[docs/excel/custom-functions-changelog.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/excel/custom-functions-changelog.md) | :white_check_mark:Succeeded | |
[docs/images/Batch1.svg](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/images/Batch1.svg) | :white_check_mark:Succeeded | |
[docs/images/Batch2.svg](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/images/Batch2.svg) | :white_check_mark:Succeeded | |
[docs/images/Batch3.svg](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/images/Batch3.svg) | :white_check_mark:Succeeded | |
[docs/images/Batch4.svg](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/images/Batch4.svg) | :white_check_mark:Succeeded | |
[docs/toc.yml](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/toc.yml) | :white_check_mark:Succeeded | |

<a name="docs%2Fexcel%2Fcustom-functions-batching.md"></a>
### [docs/excel/custom-functions-batching.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/excel/custom-functions-batching.md)
 - **Line 12**: **[Warning]** Invalid file link:(~/tutorials/).

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C1%5C30%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201901302218533534-712%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=cQbaaOoJN1yannBZGIVO2o1znu5uY6q7oiwHufjMhPc%3D&st=2019-01-30T22%3A15%3A15Z&se=2019-03-02T22%3A20%3A15Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: docs/excel/custom-functions-batching.md (hunk 4 above)
Comment (lindalu-MSFT) [reply docs/excel/custom-functions-batching.md]: Thanks!

**Comment by lindalu-MSFT** [comment]
@mscharlock It looks like you are missing the file name in the URL in line 12 so you're getting the invalid file link error. It should be ~/tutorials/excel-tutorial-create-custom-functions.md

**Comment by o365devx** [comment]
OPS Build status updates of commit _[a009b0b](https://github.com/OfficeDev/office-js-docs-pr/pull/712/commits/a009b0bc444f8e75a38bf31b9945305b367a4e5b)_: 

:clock10: **Preparing**: average preparing time is _44 sec(s)_

**Comment by o365devx** [comment]
OPS Build status updates of commit _[a009b0b](https://github.com/OfficeDev/office-js-docs-pr/pull/712/commits/a009b0bc444f8e75a38bf31b9945305b367a4e5b)_: 

:clock10: **Preparing**: average preparing time is _44 sec(s)_

**Comment by o365devx** [comment]
OPS Build status updates of commit _[a009b0b](https://github.com/OfficeDev/office-js-docs-pr/pull/712/commits/a009b0bc444f8e75a38bf31b9945305b367a4e5b)_: 

:clock10: **Incremental building**: average incremental building time is _29 sec(s)_

**Comment by o365devx** [comment]
OPS Build status updates of commit _[a009b0b](https://github.com/OfficeDev/office-js-docs-pr/pull/712/commits/a009b0bc444f8e75a38bf31b9945305b367a4e5b)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/excel/custom-functions-batching.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/excel/custom-functions-batching.md) | :white_check_mark:Succeeded | |
[docs/excel/custom-functions-changelog.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/excel/custom-functions-changelog.md) | :white_check_mark:Succeeded | |
[docs/images/Batch1.svg](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/images/Batch1.svg) | :white_check_mark:Succeeded | |
[docs/images/Batch2.svg](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/images/Batch2.svg) | :white_check_mark:Succeeded | |
[docs/images/Batch3.svg](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/images/Batch3.svg) | :white_check_mark:Succeeded | |
[docs/images/Batch4.svg](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/images/Batch4.svg) | :white_check_mark:Succeeded | |
[docs/toc.yml](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-addbatchtopic/docs/toc.yml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C1%5C30%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201901302314047642-712%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=cMakQoCpV24Iynq%2FDMIIM4KMIgzdJoae2wviAhdHqns%3D&st=2019-01-30T23%3A10%3A27Z&se=2019-03-02T23%3A15%3A27Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Pull request merged**